benchmark/scenarios/write_request_load.py

```diff
@@ -0,0 +1,349 @@
+# Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
+"""
+Read request load scenario for the control service benchmarks.
+"""
+from collections import deque
+from itertools import repeat
+
+from zope.interface import implementer
+from eliot import start_action, write_failure, Message
+from eliot.twisted import DeferredContext
+
+from twisted.internet.defer import CancelledError, Deferred
+from twisted.internet.task import LoopingCall
+
+from flocker.common import loop_until, timeout
+
+from .._interfaces import IScenario
+
+DEFAULT_SAMPLE_SIZE = 5
+
+
+class WRateMeasurer(object):
+    # XXX make it a common class for Read and Write scenarios.
+    # Note that the docstrings are not up-to-date because they are being
+    # updated in the Read scenario
+
+    """
+    Measures the rate of requests in requests per second.
+
+    :ivar sample_size: The number of samples to collect.
+    """
+
+    def __init__(self, sample_size=DEFAULT_SAMPLE_SIZE):
+        self.sample_size = sample_size
+        self._counts = deque([0] * sample_size, sample_size)
+        self._sent = 0
+        self._received = 0
+        self._errors = 0
+        self._rate = 0
+
+    def request_sent(self):
+        """
+        Increase the number of sent requests.
+        """
+        self._sent += 1
+
+    def response_received(self, ignored):
+        """
+        Increase the number of received requests.
+
+        :param ignored: The result of a callback. This parameter is
+            not used.
+        """
+        self._received += 1
+
+    def request_failed(self, ignored):
+        """
+        Increase the error count for failed requests.
+
+        :param ignored: The result of a callback. This parameter is
+            not used.
+        """
+        self._errors += 1
+
+    def update_rate(self):
+        """
+        Update the current rate and record a new sample.
+        """
+        self._rate = (
+            (self._received - self._counts[0]) / float(self.sample_size)
+        )
+        self._counts.append(self._received)
+
+    def outstanding(self):
+        """
+        Return the number of outstanding requests.
+        """
+        return self._sent - self._received - self._errors
+
+    def rate(self):
+        return self._rate
+
+
+# XXX make this excetions common for read and write scenarios and remove
+# the initial W preceding the name
+class WRequestRateTooLow(Exception):
+    """
+    The request rate dropped below a threshold.
+    """
+
+
+class WRequestRateNotReached(Exception):
+    """
+    The request rate did not reach the target level.
+    """
+
+
+class WRequestOverload(Exception):
+    """
+    There are too many outstanding requests.
+    """
+
+
+class WDataseCreationTimeout(Exception):
```

Comment:
Add "t" to the end of "Datase" and make sure you change it in other files in which it appears.